driver/internal.go

```diff
@@ -83,7 +88,53 @@ func configAuth(config *config.Driver) (*clientcredentials.Config, context.Conte
 		tokenURL = TokenURLOverride
 	}
 
-	oCfg := &clientcredentials.Config{TokenURL: tokenURL, ClientID: appID, ClientSecret: appSecret, AuthStyle: oauth2.AuthStyleInParams}
+	ctxClient := context.WithValue(context.Background(), oauth2.HTTPClient, &http.Client{Transport: tr, Timeout: tokenRequestTimeout})
+
+	var ts oauth2.TokenSource
+	var client *http.Client
+
+	// use access-token authentication if it's set,
+	// use client_id, client_secret otherwise
+	if token != "" {
+		t := &oauth2.Token{AccessToken: token}
+		ocfg1 := &oauth2.Config{Endpoint: oauth2.Endpoint{TokenURL: tokenURL}}
+		client = ocfg1.Client(ctxClient, t)
+		ts = ocfg1.TokenSource(ctxClient, t)
+	} else if appID != "" || appSecret != "" {
+		oCfg := &clientcredentials.Config{TokenURL: tokenURL, ClientID: appID, ClientSecret: appSecret, AuthStyle: oauth2.AuthStyleInParams}
+		client = oCfg.Client(ctxClient)
+		ts = oCfg.TokenSource(ctxClient)
+	}
+
+	// token source is configured for GRPC token retrieval.
+	// client is configured for HTTP token retrieval
+	return ts, client
+}
+
+func ToString(s *string) string {
```

Comment:
nit: you can name these functions something like `FromPtrToString` as ToString or ToBytes here is more generic.